crawl-ref/source/traps.cc

```diff
@@ -1320,7 +1320,7 @@ bool is_valid_shaft_effect_level()
     // Don't shaft the player when we can't, and also when it would be into a
     // dangerous end.
     return is_valid_shaft_level()
-           && !(testbits(branch.branch_flags, brflag::dangerous_end)
```

Comment:
Looking now at the git blame, I see that someone added in the use of the testbits function here with the idea that this use of bitwise & in a return statement wouldn't compile? (6e6263181c) I don't really know what they were thinking; it seemed to compile fine for me and now seems to be passing Travis' tests. But I thought I'd flag it since I'm not sure what they had in mind.

Reply:
In this instance I believe it compiles fine, because it's converted to bool before the return statement acts. Return statements of the form `return a & b`, where `b` is of an enum class type, in a boolean function/method however don't compile, and this particular change was caught up in those sorts of fixes. What exactly is the issue with using `testbits` though?

FWIW I now use `static_cast`s to bool for return statement compile errors when converting enums.

Reply:
Oh, great!---thanks for explaining!

Why not use testbits here? Well of course I will defer to you more knowledgeable folks, but fwiw here was my reasoning:

- in the case we care about, where the test field is has a single bit turned on, using testbits doesn't change the behavior compared to a & b (or rather static_cast<bool> (a & b))
- using & to query a bit is common and readability isn't helped by calling testbits in addition
- in general, if some verbosity doesn't change the behavior or make the code more readable, it should be dropped

Reply:
`&` is fine here.